partition_processor.go

```diff
@@ -297,50 +294,58 @@ func (pp *PartitionProcessor) Start(ctx context.Context) error {
 	return nil
 }
 
+func (pp *PartitionProcessor) stopping() <-chan struct{} {
+	return pp.state.WaitForStateMin(PPStateStopping)
+}
+
 // Stop stops the partition processor
 func (pp *PartitionProcessor) Stop() error {
 	pp.log.Debugf("Stopping")
 	defer pp.log.Debugf("... Stopping done")
 	pp.state.SetState(PPStateStopping)
-	defer pp.state.SetState(PPStateIdle)
-	errs := new(multierr.Errors)
+	defer pp.state.SetState(PPStateStopped)
 
 	if pp.cancelRunnerGroup != nil {
 		pp.cancelRunnerGroup()
 	}
-	if pp.runnerGroup != nil {
-		errs.Collect(<-pp.Errors())
-	}
 
 	// stop the stats updating/serving loop
 	pp.cancelStatsLoop()
 
-	errg, _ := multierr.NewErrGroup(context.Background())
+	// wait for the runner to be done
+	runningErrs := multierror.Append(pp.runnerGroup.Wait().ErrorOrNil())
+
+	// close all the tables
+	stopErrg, _ := multierr.NewErrGroup(context.Background())
 	for _, join := range pp.joins {
 		join := join
-		errg.Go(func() error {
+		stopErrg.Go(func() error {
 			return join.Close()
 		})
 	}
 
+	// close processor table, if there is one
 	if pp.table != nil {
-		errg.Go(func() error {
+		stopErrg.Go(func() error {
 			return pp.table.Close()
 		})
 	}
-	errs.Collect(errg.Wait().NilOrError())
 
-	return errs.NilOrError()
+	// return stopping errors and running errors
+	return multierror.Append(stopErrg.Wait().ErrorOrNil(), runningErrs).ErrorOrNil()
 }
 
 func (pp *PartitionProcessor) run(ctx context.Context) (rerr error) {
 	pp.log.Debugf("starting")
 	defer pp.log.Debugf("stopped")
 
-	errs := new(multierr.Errors)
+	// protect the errors-collection
+	var mutexErr sync.Mutex
+
 	defer func() {
```

Comment:
The mutex is used a lot in this long func. You could add a `setOrAddError` func that has uses the mutex exclusively and use this func in all places where you want to set or add errors.

This func could then also do a nil check and append if not nil to ensure no errors are lost.